File: backend/app/services/nvidia.py
"""NVIDIA NIM chat integration (OpenAI-compatible streaming API).

The API key lives only on the backend. The frontend never sees it.
"""

from __future__ import annotations

import json
from collections.abc import AsyncGenerator

import httpx

from ..config import settings
from .context import SYSTEM_PROMPT


def is_configured() -> bool:
    return bool(settings.nvidia_nim_api_key)


async def stream_chat(messages: list[dict[str, str]]) -> AsyncGenerator[str, None]:
    """Stream assistant tokens from NVIDIA NIM.

    `messages` is the conversation history ([{role, content}, ...]).
    Yields text deltas. Raises httpx.HTTPError on transport failure so the
    caller can fall back gracefully.
    """
    payload = {
        "model": settings.nvidia_nim_model,
        "messages": [{"role": "system", "content": SYSTEM_PROMPT}, *messages],
        "temperature": 0.3,
        "top_p": 0.9,
        "max_tokens": 1024,
        "stream": True,
    }
    headers = {
        "Authorization": f"Bearer {settings.nvidia_nim_api_key}",
        "Accept": "text/event-stream",
        "Content-Type": "application/json",
    }

    url = f"{settings.nvidia_nim_base_url.rstrip('/')}/chat/completions"
    timeout = httpx.Timeout(60.0, connect=10.0)

    async with httpx.AsyncClient(timeout=timeout) as client:
        async with client.stream("POST", url, json=payload, headers=headers) as resp:
            resp.raise_for_status()
            async for line in resp.aiter_lines():
                if not line or not line.startswith("data:"):
                    continue
                chunk = line[len("data:") :].strip()
                if chunk == "[DONE]":
                    break
                try:
                    parsed = json.loads(chunk)
                    delta = parsed["choices"][0]["delta"].get("content")
                    if delta:
                        yield delta
                except (json.JSONDecodeError, KeyError, IndexError):
                    continue

File: backend/src/services/nvidia.ts
/**
 * NVIDIA NIM chat integration (OpenAI-compatible streaming API).
 * The API key lives only on the backend. The frontend never sees it.
 */

import { settings } from '../config'
import type { ChatMessage } from '../types'
import { SYSTEM_PROMPT } from './context'

export function isConfigured(): boolean {
  return Boolean(settings.nvidiaNimApiKey)
}

/**
 * Stream assistant text deltas from NVIDIA NIM. `onDelta` is called for each
 * token. Throws on transport/HTTP failure so the caller can fall back.
 */
export async function streamChat(
  messages: ChatMessage[],
  onDelta: (delta: string) => void | Promise<void>,
  signal?: AbortSignal,
): Promise<void> {
  const payload = {
    model: settings.nvidiaNimModel,
    messages: [{ role: 'system', content: SYSTEM_PROMPT }, ...messages],
    temperature: 0.3,
    top_p: 0.9,
    max_tokens: 1024,
    stream: true,
  }
  const url = `${settings.nvidiaNimBaseUrl.replace(/\/$/, '')}/chat/completions`

  const controller = new AbortController()
  const timeout = setTimeout(() => controller.abort(), 60_000)
  if (signal) signal.addEventListener('abort', () => controller.abort())

  try {
    const resp = await fetch(url, {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${settings.nvidiaNimApiKey}`,
        Accept: 'text/event-stream',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(payload),
      signal: controller.signal,
    })
    if (!resp.ok || !resp.body) {
      throw new Error(`NIM responded ${resp.status}`)
    }

    const reader = resp.body.getReader()
    const decoder = new TextDecoder()
    let buffer = ''
    for (;;) {
      const { done, value } = await reader.read()
      if (done) break
      buffer += decoder.decode(value, { stream: true })
      // Server-sent events are newline-delimited.
      const lines = buffer.split('\n')
      buffer = lines.pop() ?? ''
      for (const line of lines) {
        const trimmed = line.trim()
        if (!trimmed || !trimmed.startsWith('data:')) continue
        const chunk = trimmed.slice('data:'.length).trim()
        if (chunk === '[DONE]') return
        try {
          const parsed = JSON.parse(chunk)
          const delta = parsed?.choices?.[0]?.delta?.content
          if (delta) await onDelta(delta)
        } catch {
          // Ignore malformed keep-alive/partial chunks.
        }
      }
    }
  } finally {
    clearTimeout(timeout)
  }
}
